src/halmos/cheatcodes.py

```diff
@@ -230,6 +231,45 @@ def symbolic_storage(ex, arg, sevm, stack, step_id):
     return ByteVec()  # empty return data
 
 
+def snapshot_storage(ex, arg, sevm, stack, step_id):
+    account = uint160(arg.get_word(4))
+    account_alias = sevm.resolve_address_alias(
+        ex, account, stack, step_id, allow_branching=False
+    )
+
+    if account_alias is None:
+        error_msg = f"snapshotStorage() is not allowed for a nonexistent account: {hexify(account)}"
+        raise HalmosException(error_msg)
+
+    return ByteVec(ex.storage[account_alias].digest())
+
+
+def snapshot_state(ex, arg, sevm, stack, step_id):
+    """
+    Generates a snapshot ID by hashing the current state (code, storage, and balance).
+    """
+    m = hashlib.sha3_256()
+
+    # code
+    # note: iteration order is guaranteed to be the insertion order
+    for addr, code in ex.code.items():
+        addr = addr if isinstance(addr, int) else addr.as_long()
+        m.update(int.to_bytes(addr, length=32))
+        # simply the object address is used, as code remains unchanged after deployment
+        m.update(int.to_bytes(id(code), length=32))
```

Comment:
spidey sense: I would normally avoid using the value of id in computations, since it changes from run to run